crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -243,15 +342,25 @@ impl SettlementEncoder {
         let traded_tokens: HashSet<_> = self
             .trades()
             .iter()
-            .flat_map(|trade| {
-                [
-                    trade.order.order_creation.buy_token,
-                    trade.order.order_creation.sell_token,
+            .flat_map(|order_trade| {
+                vec![
+                    order_trade.trade.order.order_creation.buy_token,
+                    order_trade.trade.order.order_creation.sell_token,
                 ]
             })
             .collect();
 
-        self.tokens.retain(|token| traded_tokens.contains(token));
+        let liquidity_traded_tokens: HashSet<_> = self
+            .liquidity_order_trades
+            .iter()
+            .flat_map(|liquidity_order_trade| {
+                vec![liquidity_order_trade.trade.order.order_creation.sell_token]
+            })
```

Comment:
nit: Using `flat_map` for a single result seems surprising. A normal map would achieve the same thing:
```rust
.map(|liquidity_order_trade| {
    liquidity_order_trade.trade.order.order_creation.sell_token
})
```